File: pcc/py_runtime/py/py_int_bitwise.py
"""pcc-Python replacement for py_runtime/src/py_int_bitwise.c."""
from pcc.extern import extern, c_abi_export, c_int64, c_ptr
from pcc.unsafe import calloc, free, load_i32, null, ptr_is_null, store_i32


py_bigint_alloc = extern("py_bigint_alloc", (c_int64,), c_ptr)


def _load_u32(obj, offset: int) -> int:
    v: int = load_i32(obj, offset)
    if v < 0:
        v = v + 4294967296
    return v


def _store_u32(obj, offset: int, value: int) -> None:
    store_i32(obj, offset, value)


def _arr_load_u32(arr, idx: int) -> int:
    return _load_u32(arr, idx * 4)


def _arr_store_u32(arr, idx: int, value: int) -> None:
    _store_u32(arr, idx * 4, value)


def _normalize(b) -> None:
    ndigits: int = load_i32(b, 20)
    while ndigits > 0 and _load_u32(b, 24 + (ndigits - 1) * 4) == 0:
        ndigits = ndigits - 1
    store_i32(b, 20, ndigits)
    if ndigits == 0:
        store_i32(b, 16, 0)
    elif load_i32(b, 16) == 0:
        store_i32(b, 16, 1)


def _encode_twos(src, out, n: int) -> None:
    ndigits: int = load_i32(src, 20)
    if load_i32(src, 16) >= 0:
        i: int = 0
        while i < ndigits:
            _arr_store_u32(out, i, _load_u32(src, 24 + i * 4))
            i = i + 1
        return

    carry: int = 0
    i = 0
    while i < n:
        d: int = 0
        if i < ndigits:
            d = _load_u32(src, 24 + i * 4)
        inv: int = 4294967295 - d
        if i == 0:
            inv = inv + 1
        inv = inv + carry
        _arr_store_u32(out, i, inv & 4294967295)
        carry = inv >> 32
        i = i + 1


def _bitop(a, b, op: int):
    na: int = load_i32(a, 20)
    nb: int = load_i32(b, 20)
    n: int = nb + 1
    if na > nb:
        n = na + 1

    sa = calloc(n, 4)
    sb = calloc(n, 4)
    if ptr_is_null(sa) or ptr_is_null(sb):
        free(sa)
        free(sb)
        return null()

    _encode_twos(a, sa, n)
    _encode_twos(b, sb, n)

    sr = calloc(n, 4)
    if ptr_is_null(sr):
        free(sa)
        free(sb)
        return null()

    i: int = 0
    while i < n:
        av: int = _arr_load_u32(sa, i)
        bv: int = _arr_load_u32(sb, i)
        rv: int = 0
        if op == 0:
            rv = av & bv
        elif op == 1:
            rv = av | bv
        else:
            rv = av ^ bv
        _arr_store_u32(sr, i, rv)
        i = i + 1

    free(sa)
    free(sb)

    r = py_bigint_alloc(n)
    if ptr_is_null(r):
        free(sr)
        return r

    result_negative: bool = (_arr_load_u32(sr, n - 1) & 2147483648) != 0
    if not result_negative:
        i = 0
        while i < n:
            _store_u32(r, 24 + i * 4, _arr_load_u32(sr, i))
            i = i + 1
        store_i32(r, 16, 1)
    else:
        carry: int = 1
        i = 0
        while i < n:
            inv = 4294967295 - _arr_load_u32(sr, i) + carry
            _store_u32(r, 24 + i * 4, inv & 4294967295)
            carry = inv >> 32
            i = i + 1
        store_i32(r, 16, -1)

    free(sr)
    _normalize(r)
    return r


@c_abi_export("py_bigint_and")
def py_bigint_and(a, b):
    return _bitop(a, b, 0)


@c_abi_export("py_bigint_or")
def py_bigint_or(a, b):
    return _bitop(a, b, 1)


@c_abi_export("py_bigint_xor")
def py_bigint_xor(a, b):
    return _bitop(a, b, 2)

File: pcc/py_runtime/src/py_int_bitwise.c
/* pcc/py_runtime/src/py_int_bitwise.c
 *
 * Bignum bitwise helpers split from py_int.c for independent
 * pcc-Python replacement.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <stdint.h>

static void py_bigint_normalize(PyIntObject *b) {
    while (b->ndigits > 0 && b->digits[b->ndigits - 1] == 0) {
        b->ndigits--;
    }
    if (b->ndigits == 0) b->sign = 0;
    else if (b->sign == 0) b->sign = 1;
}

enum BitOp { BIT_AND, BIT_OR, BIT_XOR };

static PyIntObject *bigint_bitop(const PyIntObject *a, const PyIntObject *b,
                                 enum BitOp op) {
    int32_t na = a->ndigits, nb = b->ndigits;
    int32_t n = (na > nb ? na : nb) + 1;    /* extra sign-extension digit */
    uint32_t *sa = (uint32_t *)calloc((size_t)n, sizeof(uint32_t));
    uint32_t *sb = (uint32_t *)calloc((size_t)n, sizeof(uint32_t));
    if (!sa || !sb) { free(sa); free(sb); return NULL; }

    /* Fill sa with two's-complement representation of a. */
    if (a->sign >= 0) {
        for (int32_t i = 0; i < na; i++) sa[i] = a->digits[i];
    } else {
        uint64_t borrow = 0;
        for (int32_t i = 0; i < n; i++) {
            uint32_t d = (i < na) ? a->digits[i] : 0;
            /* two's complement: ~d + (i==0 ? 1 : 0) with carry */
            uint64_t inv = (uint32_t)~d;
            if (i == 0) inv += 1;
            inv += borrow;
            sa[i] = (uint32_t)(inv & 0xFFFFFFFFu);
            borrow = inv >> 32;
        }
    }
    if (b->sign >= 0) {
        for (int32_t i = 0; i < nb; i++) sb[i] = b->digits[i];
    } else {
        uint64_t borrow = 0;
        for (int32_t i = 0; i < n; i++) {
            uint32_t d = (i < nb) ? b->digits[i] : 0;
            uint64_t inv = (uint32_t)~d;
            if (i == 0) inv += 1;
            inv += borrow;
            sb[i] = (uint32_t)(inv & 0xFFFFFFFFu);
            borrow = inv >> 32;
        }
    }

    uint32_t *sr = (uint32_t *)calloc((size_t)n, sizeof(uint32_t));
    if (!sr) { free(sa); free(sb); return NULL; }

    for (int32_t i = 0; i < n; i++) {
        switch (op) {
            case BIT_AND: sr[i] = sa[i] & sb[i]; break;
            case BIT_OR:  sr[i] = sa[i] | sb[i]; break;
            case BIT_XOR: sr[i] = sa[i] ^ sb[i]; break;
        }
    }
    free(sa);
    free(sb);

    /* Decode result: sign is determined by the top bit of sr[n-1]. */
    int result_sign = (sr[n - 1] & 0x80000000u) ? -1 : 1;
    PyIntObject *r = py_bigint_alloc(n);
    if (r == NULL) { free(sr); return NULL; }

    if (result_sign >= 0) {
        for (int32_t i = 0; i < n; i++) r->digits[i] = sr[i];
        r->sign = 1;
    } else {
        /* magnitude = 2^(32n) - sr == ~sr + 1 */
        uint64_t carry = 1;
        for (int32_t i = 0; i < n; i++) {
            uint64_t inv = (uint32_t)~sr[i];
            inv += carry;
            r->digits[i] = (uint32_t)(inv & 0xFFFFFFFFu);
            carry = inv >> 32;
        }
        r->sign = -1;
    }
    free(sr);
    py_bigint_normalize(r);
    return r;
}

PyIntObject *py_bigint_and(const PyIntObject *a, const PyIntObject *b) { return bigint_bitop(a, b, BIT_AND); }
PyIntObject *py_bigint_or (const PyIntObject *a, const PyIntObject *b) { return bigint_bitop(a, b, BIT_OR);  }
PyIntObject *py_bigint_xor(const PyIntObject *a, const PyIntObject *b) { return bigint_bitop(a, b, BIT_XOR); }
